sympy/combinatorics/groups.py

```diff
@@ -0,0 +1,157 @@
+class FreeGroup(Group):
+    r"""
+    Called  with  a  positive  integer rank, FreeGroup
+    returns a free group on rank generators. If optional
+    argument name is given then the  generators are printed
+    as name1, name2 etc., that is, each name is the
+    concatenation of the string name and an integer from 1
+    to range. The default for name is the string "f".
+    """
+    is_group = True
+
+    def __new__(cls, *args, **kwargs):
+        if not(len(args) == 1 or isinstance(args[0], int)):
+            raise ValueError("")
+
+        obj = Basic.__new__(cls, *args)
+        obj._generators = args
+        obj._order = S.Infinity
+        obj._center = []
+        obj._degree = obj._generators
+        obj._is_abelian = False
+
+        return obj
+
+    def __init__(self, *args):
+        if isinstance(args[0], int):
+            self.gens_assign(*args)
+
+    def gens_assign(self, *args):
+        f = var('f0:%s' %args)
+
+    def __repr__(self):
+        str_form = '<free group on the generators[ '
+        for i in range(len(self)):
+            str_form += 'f%s, ' %i
+        str_form = str_form[:-2]
+        str_form += ' ]>'
+        return str_form
+
+    def __len__(self):
+        if isinstance(self.args[0], int):
+            return self.args[0]
+
+    def __getitem__(self):
+        return self._generators[i]
+
+
+class Group(Basic):
+    pass
+
+
+class GroupElem(Basic):
+    """
+    Group element definition
+
+    """
+
+    def __init__(self, elem, group):
+        if not isinstance(group, Group):
+            raise TypeError("group is not a Group")
+        if not elem in group.Set:
+            raise TypeError("elem is not an element of group")
+        self.elem = elem
+        self.group = group
+
+    def __str__(self):
+        return str(self.elem)
+
+    def __eq__(self, other):
+        """
+        Two GroupElems are equal if they represent the same element,
+        regardless of the Groups they belong to
+        """
+
+        if not isinstance(other, GroupElem):
+            raise TypeError("other is not a GroupElem")
+        return self.elem == other.elem
+
+    def __ne__(self, other):
+        return not self == other
+
+    def __hash__(self):
+        return hash(self.elem)
```

Comment:
If it's possible to store the state of this object in its args instead of defining `__eq__` and `__hash__` that is highly preferable. Remember that == in sympy does not have to represent mathematical equality.